doc/source/data/custom-datasource.rst

```diff
@@ -0,0 +1,105 @@
+.. _custom_datasources:
+
+==================
+Custom Datasources
+==================
+
+Ray Datasets supports multiple ways to :ref:`create a dataset <creating_datasets>`, 
+allowing you to easily ingest data of common formats from popular sources. However, if the 
+datasource you want to read from is not in the built-in list, don't worry, you can implement 
+a custom one for your use case. In this guide, we will walk you through how to build 
+your own custom datasource, using MongoDB as an example.
+
+A custom datasource is an implementation of :class:`~ray.data.Datasource`. In the 
+example here, let's call it ``MongoDatasource``. At a high level, it will have two 
+core parts to build out: read support with :meth:`create_reader() <ray.data.Datasource.create_reader>` 
+and write support with :meth:`do_write() <ray.data.Datasource.do_write>`.
+
+Here are the key design choices we will make in this guide:
+
+-  **MongoDB connector**: We will use `PyMongo <https://pymongo.readthedocs.io/en/stable/>`__ to connect to MongDB.
```

Comment:
Use the present tense.
Use `PyMongo <https://pymongo.readthedocs.io/en/stable/>`__ to connect to MongDB.

Reply:
Done here and other comments.